### PR TITLE
[Disco] Integrate RCCL

### PR DESCRIPTION
**Pull request by LeshengJin**
This PR integrates RCCL for amd multi-GPU parallelism.

### CODE REVIEW

File: src/runtime/disco/ccl/utils.h [hunk 1]
```diff
@@ -19,22 +19,21 @@
...
-namespace nccl {
+namespace ccl {
 
-#define NCCL_CALL(cmd)                                       \
-  do {                                                       \
-    ncclResult_t r = cmd;                                    \
-    if (r != ncclSuccess) {                                  \
-      LOG(FATAL) << "NCCLErrror: " << ncclGetErrorString(r); \
-    }                                                        \
+#define NCCL_CALL(cmd)                                             \
```
Comment (LeshengJin) [inline src/runtime/disco/ccl/utils.h:31]:
```suggestion
#define CCL_CALL(cmd)                                             \
```
Do you think it is necessary to do a renaming of functions in `utils.h`?

File: src/runtime/disco/ccl/utils.h [hunk 2]
```diff
@@ -19,22 +19,21 @@
...
+  do {                                                             \
+    ncclResult_t r = cmd;                                          \
+    if (r != ncclSuccess) {                                        \
+      LOG(FATAL) << CCLString "Errror: " << ncclGetErrorString(r); \
+    }                                                              \
   } while (0)
 
 inline ncclDataType_t AsNCCLDataType(runtime::DataType dtype) {
```
Comment (LeshengJin) [inline src/runtime/disco/ccl/utils.h:39]:
```suggestion
inline ncclDataType_t AsCCLDataType(runtime::DataType dtype) {
```

File: tests/python/disco/test_rccl.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (junrushao) [inline tests/python/disco/test_rccl.py:1]: you may parameterize the original `test_nccl.py`

File: tests/python/disco/test_ccl.py [hunk 4]
```diff
@@ -28,22 +28,51 @@
...
-
-
-@pytest.mark.parametrize("session_kind", _all_session_kinds)
-def test_init(session_kind):
+_ccl = get_global_func("runtime.disco.compiled_ccl")()
+_test_inputs = [(sess_kind, _ccl) for sess_kind in _all_session_kinds]
+
+
+def create_device_target(ccl):
+    if ccl == "nccl":
```
Comment (junrushao) [inline tests/python/disco/test_ccl.py:39]: can you use `target.from_device("cuda:0")` instead?

File: tests/python/disco/test_ccl.py [hunk 5]
```diff
@@ -28,22 +28,51 @@
...
 from tvm.script import relax as R
+from tvm import get_global_func
 
 _all_session_kinds = [di.ThreadedSession, di.ProcessSession]
-
-
-@pytest.mark.parametrize("session_kind", _all_session_kinds)
-def test_init(session_kind):
+_ccl = get_global_func("runtime.disco.compiled_ccl")()
+_test_inputs = [(sess_kind, _ccl) for sess_kind in _all_session_kinds]
```
Comment (junrushao) [inline tests/python/disco/test_ccl.py:35]: we should test "sess_kind x ccl" instead

File: src/runtime/disco/ccl/ccl.cc [hunk 6]
```diff
@@ -16,9 +16,48 @@
...
+
+#define runtimeCall CUDA_CALL
+#define SetDevice cudaSetDevice
+#define StreamCreate cudaStreamCreate
+#define StreamSync cudaStreamSynchronize
+#define StreamDestroy cudaStreamDestroy
+#define CommDestroy ncclCommDestroy
+#define RuntimeDeviceType DLDeviceType::kDLCUDA
+#define runtimeString "cuda"
+#define CCLString "nccl"
```
Comment (junrushao) [inline src/runtime/disco/ccl/ccl.cc:39]: if possible, avoid `#define` for constant values, and we should follow google c++ naming style:

```suggestion
#define TVM_DISCO_DEVICE_CALL CUDA_CALL
#define TVM_DISCO_DEVICE_SET_DEVICE cudaSetDevice
#define TVM_DISCO_DEVICE_STREAM_CREATE cudaStreamCreate
#define TVM_DISCO_DEVICE_STREAM_SYNC cudaStreamSynchronize
#define TVM_DISCO_DEVICE_STREAM_DESTROY cudaStreamDestroy
#define TVM_DISCO_DEVICE_NAME "cuda"
#define TVM_DISCO_CCL_DESTROY ncclCommDestroy
#define TVM_DISCO_CCL_NAME "nccl"
const constexpr DLDeviceType = DLDeviceType::kDLCUDA;
```

**Review by jinhongyii — APPROVED** [review]

**Pull request merged**